### ISSUE TITLE
Business case matrix should update when (for example) assets are removed

### ISSUE THREAD

**Issue by ChaelKruip** [quintel/etmoses#1468]
Now you first have to recalculate the yearly business case before the business case matrix is updated which is not very clear to the user.

**Comment by grdw** [comment]
I don't understand what you mean? Which matrix? The matrix on the edit page?

**Comment by ChaelKruip** [comment]
> Which matrix? The matrix on the edit page?

Yes.

**Comment by grdw** [comment]
What would be cool to do is, when there's a change to either the `heat_asset_list`, `heat_source_list`, `gas_asset_list`, `topology` or `technology_matrix` detect the change and automatically recalculate the business case. 

The thing is, come to think of it, that `heat_asset_list`, `heat_source_list`, `gas_asset_list`, `topology` should only need to recalculate the depreciation costs of the business case. Changes in the `technology_matrix` changes the load and the depreciation costs on the network and therefor the business case as a whole. 

It would be nice if we could separate the two calculations before thinking about recalculating whole business cases just because someone removes a row in for example, the `heat_source_list`.

**Comment by ChaelKruip** [comment]
> The thing is, come to think of it, that heat_asset_list, heat_source_list, gas_asset_list, topology should only need to recalculate the depreciation costs of the business case. Changes in the technology_matrix changes the load and the depreciation costs on the network and therefor the business case as a whole.

Unfortunately, it is not that easy. The energy flows can change by editing of heat_asset_list, heat_source_list, gas_asset_list and topology as well resulting in a different outcome of the market model transactions.

> It would be nice if we could separate the two calculations before thinking about recalculating whole business cases just because someone removes a row in for example, the heat_source_list.

Yes. But because if is not clear how to do that I propose to trigger recalculation (in the background) with every change that is saved.

**Comment by antw** [comment]
> Unfortunately, it is not that easy. The energy flows can change by editing of heat_asset_list, heat_source_list, gas_asset_list and topology as well resulting in a different outcome of the market model transactions.

Only the heat source list can affect energy flows within the network calculation.

**Comment by ChaelKruip** [comment]
> Only the heat source list can affect energy flows within the network calculation.

Right! I forgot that we currently don't include the impact of constraints in the flows. Good point @antw 👍

**Comment by grdw** [comment]
> Yes. But because if is not clear how to do that I propose to trigger recalculation (in the background) with every change that is saved.

I think this needs to be like this:

-> Make change to for example heat source list 
-> Check for pending business case jobs
-> If so, don't do anything, if not recalculate business case


Todo:
- [ ] Show spinner in header of business case (edit)
- [ ] Redraw the business case matrix when the business case calcualation is done (`Poller`).

**Comment by grdw** [comment]
I am currently thinking about this. All updates go through a `remote` js request. So technically there should be a javascript script which extends `FinanceTable` (I'd probably rename that file to `BusinessCaseMatrix`). 

Can't we recalculate the business case and redraw the matrix based on every clicked 'Save'.  (Not "Save all and view LES" since you move of the Edit page when you do that)? To recalculate on every change sounds a little like something that is part of #776 .

**Comment by grdw** [comment]
Ping @ChaelKruip and @antw

**Comment by ChaelKruip** [comment]
> Can't we recalculate the business case and redraw the matrix based on every clicked 'Save'. 

👍

**Comment by antw** [comment]
I don't think it needs to be recalculated after every change; only if the visitor tries to look at the matrix and the cached copy is stale.

I feel like the "Edit → Business Case" page isn't as necessary as it used to be, since the "proper" (?) business case page shows a breakdown of costs:

![screen shot 2016-10-31 at 11 38 50](https://cloud.githubusercontent.com/assets/4383/19853046/bdb0a3f8-9f5e-11e6-801d-bc5888013259.png)

If a matrix is necessary, I think it should be moved to the proper business case page.

**Comment by ChaelKruip** [comment]
>I don't think it needs to be recalculated after every change; only if the visitor tries to look at the matrix and the cached copy is stale.

Ans also, if it is used as a reference for a comparison in another LES.

>I feel like the "Edit → Business Case" page isn't as necessary as it used to be

I think I agree. The only functionality that is not yet included in the current business case output page is the 'free-form/correction' row.

**Comment by grdw** [comment]
> I think I agree. The only functionality that is not yet included in the current business case output page is the 'free-form/correction' row.

Yes. We're bit dragging of the main point of the issue here 😆. But we could make the "freeform" or "correction" as it is called on the business case page editable on the 'edit' page. 

So to explain in pictures:

This would be the business case page:

![screen shot 2016-10-31 at 14 57 37](https://cloud.githubusercontent.com/assets/2676542/19856580/63769368-9f7a-11e6-946c-ffd05188acac.png)

This would be the edit page:

![screen shot 2016-10-31 at 15 00 27](https://cloud.githubusercontent.com/assets/2676542/19856685/d10731a8-9f7a-11e6-864e-b540ed08003d.png)

**Comment by antw** [comment]
> Yes. We're bit dragging of the main point of the issue here 😆.

Well, what I mean is we're only discussing the need to recalculate the business case for the "matrix" page because it shows the costs for each stakeholder. If it no longer shows the matrix (and therefore no longer shows any costs except the "correction" field), then the business case does not need to be recalculated.

I don't see why the "Edit → Business Case" page needs to exist anymore. The "correction" field could be moved to the show business case page (like the "edit" screenshot above) and the "edit" page could be removed entirely.

**Comment by grdw** [comment]
> [..] and the "edit" page could be removed entirely.

Yes, it's a bit obsolete.

**Comment by ChaelKruip** [comment]
> Well, what I mean is we're only discussing the need to recalculate the business case for the "matrix" page because it shows the costs for each stakeholder. If it no longer shows the matrix (and therefore no longer shows any costs except the "correction" field), then the business case does not need to be recalculated.

But also, if a business case is used as a reference for a comparison in another LES.
